mlflow/utils/_capture_modules.py

```diff
@@ -82,18 +82,29 @@ def main():
     # Mirror `sys.path` of the parent process
     sys.path = json.loads(args.sys_path)
 
-    # If `model_path` is a directory containing an `MLmodel` file, get the model data path from it
+    cap_cm = _CaptureImportedModules()
+
+    # If `model_path` refers to an MLflow model directory, load the model using
+    # `mlflow.pyfunc.load_model`
     if os.path.isdir(model_path) and MLMODEL_FILE_NAME in os.listdir(model_path):
-        conf = mlflow.models.Model.load(model_path).flavors.get(flavor)
-        model_path = os.path.join(model_path, conf[DATA]) if (DATA in conf) else model_path
+        pyfunc_conf = Model.load(model_path).flavors.get(mlflow.pyfunc.FLAVOR_NAME)
+        loader_module = importlib.import_module(pyfunc_conf[MAIN])
+        original = loader_module._load_pyfunc
+
+        @functools.wraps(original)
+        def _load_pyfunc_patch(*args, **kwargs):
+            with cap_cm:
+                return original(*args, **kwargs)
 
-    # Load the model and capture modules imported during the loading procedure
-    flavor_module = getattr(mlflow, flavor)
-    with _CaptureImportedModules() as cap:
-        flavor_module._load_pyfunc(model_path)
+        loader_module._load_pyfunc = _load_pyfunc_patch
+        mlflow.pyfunc.load_model(model_path)
+    # Otherwise, load the model using `mlflow.<flavor>._load_pyfunc`
+    else:
+        with cap_cm:
+            importlib.import_module(f"mlflow.{flavor}")._load_pyfunc(model_path)
```

Comment:
We use this branch for scikit-learn and spacy models. In scikit-learn, we exclude pyfunc flavor if the model to log doesn't implement a predict method (we do something similar for spacy).

Reply:
Can we include this in the inline comment?

Reply:
Sure!